src/components/verification-grid/verification-grid.ts

```diff
@@ -292,6 +338,194 @@ export class VerificationGridComponent extends AbstractComponent(LitElement) {
+    const usableContainerSize: Size = {
+      width: containerSize.width - totalGapWidth,
+      height: containerSize.height - totalGapHeight,
+    };
+
+    const width = usableContainerSize.width / gridShape.columns;
+    const height = usableContainerSize.height / gridShape.rows;
+
+    return { width, height };
+  }
+
+  private willFitTileSize(containerSize: Size, gridShape: GridShape): boolean {
+    const minimumTileWidth: Pixel = this.minimumGridSize.width;
+    const minimumTileHeight: Pixel = this.minimumGridSize.height;
+
+    const proposedTileSize = this.tileSize(containerSize, gridShape);
+
+    const meetsMinimumWidth = proposedTileSize.width >= minimumTileWidth;
+    const meetsMinimumHeight = proposedTileSize.height >= minimumTileHeight;
+
+    return meetsMinimumHeight && meetsMinimumWidth;
+  }
+
+  private gridAspectRatioSimilarity(containerSize: Size, gridShape: GridShape): number {
+    const targetAspectRatio = containerSize.width / containerSize.height;
+    const candidateSizeAspectRatio = gridShape.columns / gridShape.rows;
+
+    const distance = Math.abs(targetAspectRatio - candidateSizeAspectRatio);
+    const maxRatio = Math.max(targetAspectRatio, candidateSizeAspectRatio);
+    const similarity = 1 - distance / maxRatio;
+
+    return similarity;
+  }
+
+  private computeGridShape(target: number): GridShape {
+    if (target === 1) {
+      return { columns: 1, rows: 1 };
+    } else if (target === 4) {
+      // TODO: Find out why 2x2 is not being preferred in our current algorithm
+      // and fix it so that we don't have to use this hack
+      //
+      // our current implementation works well for most grid sizes
+      // however, for a grid target of 4, it will prefer a 4x1 grid over a 2x2
+      // this is because 2x2 does not meet our aspect ratio threshold
+      // this is expected because we don't usually want a 1:1 aspect ratio
+      // for a 16:9 screen
+      // however, in the case of a target of 4, we want a 2x2 grid
+      // I've used this hack to force a 2x2 grid for a target of 4
+      // when the grid containers aspect ratio is close to 16:9
+      const screenSize = window.screen;
+      const aspectRatio = screenSize.width / screenSize.height;
+
+      // we allow for a bit of deviation from the exact 16:9 aspect ratio so
+      // similar aspect ratios also get the nicer grid size for a target of 4
+      const matchingAspectRatio = 16 / 9;
+      const matchingDeviation = 0.2;
+
+      if (Math.abs(aspectRatio - matchingAspectRatio) <= matchingDeviation) {
+        return { columns: 2, rows: 2 };
+      }
+    }
+
+    const containerSize = this.gridContainer.getBoundingClientRect();
+    let refinedTarget = target;
+    const testedTargets: number[] = [];
+
+    // some numbers e.g. 9 have one factor such as 3x3 which creates a aspect
+    // ratio of 1.00
+    // using a 3x3 grid for a 16:9 screen is not ideal because it will create
+    // a lot of dead space
```

Comment:
our spectrograms stretch? do they create dead space?

Reply:
This is a bad comment, a 3x3 grid does not create "dead space" (because our spectrograms stretch)

The comment should read

> Using a 3x3 grid for a 16:9 screen is not ideal because the spectrograms will have to stretch a lot to fit. Resulting in spectrograms that have been warped and potentially lose information